File: integrations/orthanc-plugin/orthanc_stable_study.py
"""Orthanc Python plugin — fires the orchestrator webhook when a study becomes stable.

Wiring notes
------------
The Orthanc Python plugin does NOT expose a discrete ``RegisterOnStableStudyCallback``.
Lifecycle events go through ``RegisterOnChangeCallback(fn)`` where
``fn(changeType, level, resourceId)``, and STABLE_STUDY is one of the change-type
constants (see the Orthanc SDK's ``OrthancPluginChangeType`` enum). We filter on
that constant and fetch the study's tags via
``RestApiGet('/studies/{id}?requested-tags=ModalitiesInStudy')`` — the same source
of truth the Lua fallback uses, so both paths emit a byte-identical payload and the
orchestrator ingress cannot tell which mechanism fired. ``occurredAt`` is reshaped
from Orthanc's ``LastUpdate`` (DICOM ``YYYYMMDDTHHMMSS``) into RFC 3339.

Deploy
------
Load via Orthanc's Python plugin (``PythonScript`` in ``orthanc.json``). If Python
plugin packaging is awkward for a given deployment, the Lua fallback
(``orthanc_stable_study.lua``) POSTs the identical JSON body — load one or the
other, never both.

Owner: Parvati. Trigger map: ARCHITECTURE.md
"""
from __future__ import annotations

import json
import os
import urllib.parse
import urllib.request
from datetime import datetime, timezone

# The `orthanc` module is provided by the Orthanc Python plugin runtime. When this
# file is imported OUTSIDE Orthanc (e.g. from unit tests), the import fails and the
# pure helpers below stay usable. Callback registration only happens when the
# module is present, so importing the file at test time has no side effect.
try:
    import orthanc  # type: ignore[import-not-found]
except ImportError:  # pragma: no cover - only reachable in the Orthanc runtime
    orthanc = None  # type: ignore[assignment]


ORCH_WEBHOOK = os.environ.get("ORCH_WEBHOOK_URL", "http://orchestrator:8090/webhooks/orthanc")


# ---------------------------------------------------------------------------
# Pure helpers — unit-testable outside Orthanc.
# ---------------------------------------------------------------------------


def now_iso_utc() -> str:
    """RFC 3339 UTC "now" — matches the Lua fallback's ``nowIsoUtc()``. Used as
    the fall-through for ``occurredAt`` so the event stays schema-valid when a
    build's study record omits ``LastUpdate``."""
    return datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")


def to_rfc3339_utc(orthanc_dt) -> str | None:
    """Convert an Orthanc ``LastUpdate`` timestamp to RFC 3339, or None if it is
    empty/unparseable. Orthanc reports ``LastUpdate`` in the DICOM datetime shape
    ``YYYYMMDDTHHMMSS`` (UTC), which is NOT what the schema's ``format: date-time``
    (RFC 3339) requires, so we reshape it to ``YYYY-MM-DDTHH:MM:SSZ``. A value that
    is already RFC 3339 (some builds) is passed through unchanged. Mirrors the Lua
    fallback's ``toRfc3339Utc``."""
    if not isinstance(orthanc_dt, str) or not orthanc_dt:
        return None
    try:
        return datetime.strptime(orthanc_dt, "%Y%m%dT%H%M%S").strftime("%Y-%m-%dT%H:%M:%SZ")
    except ValueError:
        try:  # already RFC 3339? keep it as-is
            datetime.fromisoformat(orthanc_dt.replace("Z", "+00:00"))
            return orthanc_dt
        except ValueError:
            return None


def is_http_url(url: str) -> bool:
    """``urllib.request.urlopen`` also dereferences ``file:``/``ftp:`` schemes
    (Bandit B310 / CWE-939). ``ORCH_WEBHOOK`` is operator-configured, but
    constraining the scheme stops an accidental or tampered non-HTTP URL from
    ever being fetched. Mirrors the Lua fallback's ``isHttpUrl`` guard."""
    return urllib.parse.urlparse(url).scheme in ("http", "https")


def build_event(orthanc_study_id: str, study_record: dict | None) -> dict:
    """Assemble a schema-valid ``OrthancStableStudyEvent`` from a
    ``/studies/{id}`` REST view of the study. Field-for-field identical to the
    Lua fallback's ``buildEvent`` — the ingress cannot tell which path fired.

    Sources tags via the REST view (rather than the callback's ``resource_id``
    alone) because that is the single portable shape across every Orthanc build.
    """
    main_tags = (study_record or {}).get("MainDicomTags") or {}
    # ModalitiesInStudy is a computed study tag Orthanc only returns when explicitly
    # requested (?requested-tags=), so it lands in RequestedTags, not MainDicomTags.
    # Some builds also park AccessionNumber there.
    requested_tags = (study_record or {}).get("RequestedTags") or {}
    accession = main_tags.get("AccessionNumber") or requested_tags.get("AccessionNumber") or ""
    modality = (requested_tags.get("ModalitiesInStudy")
                or main_tags.get("ModalitiesInStudy")
                or main_tags.get("Modality") or "")
    study_uid = main_tags.get("StudyInstanceUID") or ""
    occurred_at = to_rfc3339_utc((study_record or {}).get("LastUpdate")) or now_iso_utc()
    return {
        "schemaVersion": "1.0.0",
        "eventType": "orthanc.study.stable",
        "orthancStudyId": orthanc_study_id,
        "studyInstanceUID": study_uid,
        "modality": modality,
        "accessionNumber": accession,
        "occurredAt": occurred_at,
    }


def _post(payload: dict) -> None:
    if not is_http_url(ORCH_WEBHOOK):
        raise ValueError(f"refusing non-HTTP(S) orchestrator webhook URL: {ORCH_WEBHOOK!r}")
    req = urllib.request.Request(
        ORCH_WEBHOOK,
        data=json.dumps(payload).encode(),
        headers={"Content-Type": "application/json"},
    )
    # scheme constrained to http(s) above, so urlopen cannot reach file:/ftp:
    urllib.request.urlopen(req, timeout=10)  # nosemgrep


# ---------------------------------------------------------------------------
# Orthanc callback glue — only wired when the runtime module is available.
# ---------------------------------------------------------------------------


def on_change(change_type, level, resource_id) -> None:
    """Filter for STABLE_STUDY, fetch the study record via ``RestApiGet``, build
    the event payload, POST. Any error is swallowed and logged — a downstream
    orchestrator outage must never take down the PACS."""
    if orthanc is None:  # pragma: no cover - only reachable via test-time misuse
        return
    if change_type != orthanc.ChangeType.STABLE_STUDY:
        return
    try:
        # ?requested-tags=ModalitiesInStudy makes Orthanc compute+return the study
        # modality (absent from MainDicomTags by default). Same query as the Lua path.
        study_record = json.loads(
            orthanc.RestApiGet(f"/studies/{resource_id}?requested-tags=ModalitiesInStudy")
        )
    except Exception as e:  # noqa: BLE001 - unreadable study — log and drop
        orthanc.LogError(f"OnChange(STABLE_STUDY): failed to read study {resource_id}: {e}")
        return
    payload = build_event(resource_id, study_record)
    try:
        _post(payload)
    except Exception as e:  # noqa: BLE001 - never crash the PACS on a webhook failure
        orthanc.LogError(f"orchestrator webhook failed: {e}")


if orthanc is not None:  # pragma: no cover - only in the Orthanc runtime
    orthanc.RegisterOnChangeCallback(on_change)

File: integrations/orthanc-plugin/orthanc_stable_study.lua
-- =============================================================================
-- LH-Radiology Orchestrator — OnStableStudy Lua fallback  (mitigates Risk R3)
-- =============================================================================
-- Purpose:
--   The Python plugin (orthanc_stable_study.py) is the primary path. This Lua
--   script is the drop-in fallback for deployments where packaging the Orthanc
--   Python plugin is awkward (musl base images, air-gapped hosts, minimal
--   Orthanc builds without the Python plugin baked in). It POSTs the
--   IDENTICAL webhook body to the orchestrator ingress.
--
-- IMPORTANT: load EITHER the Python plugin OR this Lua — never both, or the
-- ingress will receive duplicate stable events per study.
--
-- Contract:
--   Emits an OrthancStableStudyEvent v1.0.0
--   (contracts/events/orthanc-stable.schema.json). Field-for-field identical
--   to integrations/orthanc-plugin/orthanc_stable_study.py.
--
-- Deploy (see README.md in this directory for the full walkthrough):
--   1. Mount this file into the Orthanc container (e.g. /etc/orthanc/).
--   2. In orthanc.json:  "LuaScripts": ["/etc/orthanc/orthanc_stable_study.lua"]
--   3. Set env var ORCH_WEBHOOK_URL if the ingress isn't reachable at the
--      default http://orchestrator:8090/webhooks/orthanc.
--
-- Owner: Parvati.
-- =============================================================================

local DEFAULT_WEBHOOK = 'http://orchestrator:8090/webhooks/orthanc'

-- The Python plugin refuses non-http(s) URLs (Bandit B310 / CWE-939). Do the
-- same here so an accidental or tampered env var can never cause Orthanc to
-- dereference a file:// or ftp:// URL.
local function isHttpUrl(u)
  return type(u) == 'string'
     and (u:match('^http://') ~= nil or u:match('^https://') ~= nil)
end

local function resolveWebhook()
  local u = os.getenv('ORCH_WEBHOOK_URL')
  if u == nil or u == '' then u = DEFAULT_WEBHOOK end
  return u
end

-- RFC 3339 UTC "now". Used as a fall-through for occurredAt so the emitted
-- event is always schema-valid even if the study record is missing LastUpdate.
local function nowIsoUtc()
  return os.date('!%Y-%m-%dT%H:%M:%SZ')
end

-- Convert an Orthanc LastUpdate timestamp to RFC 3339, or nil if unparseable.
-- Orthanc reports LastUpdate as DICOM datetime 'YYYYMMDDTHHMMSS' (UTC), which is
-- NOT the schema's format: date-time (RFC 3339); reshape to 'YYYY-MM-DDTHH:MM:SSZ'.
-- A value already in RFC 3339 (has the date dashes) is passed through unchanged.
-- Mirrors the Python plugin's to_rfc3339_utc.
local function toRfc3339Utc(dt)
  if type(dt) ~= 'string' or dt == '' then return nil end
  local y, mo, d, h, mi, s = dt:match('^(%d%d%d%d)(%d%d)(%d%d)T(%d%d)(%d%d)(%d%d)$')
  if y then
    return string.format('%s-%s-%sT%s:%s:%sZ', y, mo, d, h, mi, s)
  end
  if dt:match('^%d%d%d%d%-%d%d%-%d%d') then return dt end  -- already RFC 3339
  return nil
end

-- Build the OrthancStableStudyEvent payload from a study record fetched via
-- the REST API. Sourcing tags via RestApiGet (rather than the callback's
-- tags/metadata arguments) is deliberately version-agnostic: how those args
-- are delivered to Lua callbacks varies across Orthanc builds, while the REST
-- shape is stable.
local function buildEvent(studyId, study)
  local mainTags      = (study and study.MainDicomTags) or {}
  -- ModalitiesInStudy is a computed study tag Orthanc only returns when explicitly
  -- requested (?requested-tags=), so it lands in RequestedTags, not MainDicomTags.
  -- Some builds also park AccessionNumber there.
  local requestedTags = (study and study.RequestedTags) or {}
  local accession     = mainTags.AccessionNumber or requestedTags.AccessionNumber or ''
  local modality      = requestedTags.ModalitiesInStudy or mainTags.ModalitiesInStudy
                        or mainTags.Modality or ''
  local studyUid      = mainTags.StudyInstanceUID or ''
  local occurredAt    = toRfc3339Utc(study and study.LastUpdate) or nowIsoUtc()

  return {
    schemaVersion    = '1.0.0',
    eventType        = 'orthanc.study.stable',
    orthancStudyId   = studyId,
    studyInstanceUID = studyUid,
    modality         = modality,
    accessionNumber  = accession,
    occurredAt       = occurredAt,
  }
end

-- Best-effort POST. Any error from HttpPost is swallowed by the surrounding
-- pcall in OnStableStudy — we must never raise out of the callback (would
-- fault the PACS on a downstream orchestrator outage).
local function postWebhook(url, payload)
  local body = DumpJson(payload, true)  -- true = keep string types (don't coerce)
  SetHttpTimeout(10)                    -- match the Python plugin's urlopen(timeout=10)
  -- SetHttpHeaders exists from Orthanc 1.5.x+. If your build is older, drop
  -- this line — Orthanc's Lua HTTP client will infer Content-Type from the
  -- JSON body. See README.md.
  pcall(function()
    SetHttpHeaders({ ['Content-Type'] = 'application/json' })
  end)
  HttpPost(url, body)
end

function OnStableStudy(studyId, tags, metadata)
  local webhook = resolveWebhook()
  if not isHttpUrl(webhook) then
    print('OnStableStudy: refusing non-http(s) webhook URL: ' .. tostring(webhook))
    return
  end

  local ok, study = pcall(function()
    -- ?requested-tags=ModalitiesInStudy makes Orthanc compute+return the study
    -- modality (absent from MainDicomTags by default). Same query as the Python path.
    return ParseJson(RestApiGet('/studies/' .. studyId .. '?requested-tags=ModalitiesInStudy'))
  end)
  if not ok or type(study) ~= 'table' then
    print('OnStableStudy: failed to read study ' .. tostring(studyId))
    return
  end

  local payload = buildEvent(studyId, study)

  local ok2, err = pcall(postWebhook, webhook, payload)
  if not ok2 then
    print('OnStableStudy: orchestrator webhook POST failed: ' .. tostring(err))
  end
end
